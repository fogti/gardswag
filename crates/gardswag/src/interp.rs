use gardswag_syntax::{Block, Expr};
use gardswag_varstack::VarStack;
use std::collections::BTreeMap;

#[derive(Clone, Copy, Debug, PartialEq)]
pub enum Builtin {
    Plus,
    Minus,
    Mult,
    Eq,
    Leq,
    Not,
    StdioWrite,
}

impl Builtin {
    fn argc(self) -> usize {
        match self {
            Self::Plus | Self::Minus | Self::Mult | Self::Eq | Self::Leq => 2,
            Self::Not | Self::StdioWrite => 1,
        }
    }
}

#[derive(Clone, Debug, PartialEq)]
pub enum Value<'a> {
    Unit,
    Boolean(bool),
    Integer(i32),
    PureString(String),

    Record(BTreeMap<String, Value<'a>>),

    Builtin {
        f: Builtin,
        args: Vec<Value<'a>>,
    },
    Lambda {
        argname: &'a str,
        f: &'a Expr,
        stacksave: BTreeMap<String, Value<'a>>,
    },
}

impl From<Builtin> for Value<'static> {
    fn from(x: Builtin) -> Self {
        Value::Builtin {
            f: x,
            args: Vec::new(),
        }
    }
}

pub fn run_block<'a, 's>(blk: &'a Block, stack: &'s VarStack<'s, Value<'a>>) -> Value<'a> {
    for i in &blk.stmts {
        let _ = run(i, stack);
    }
    if let Some(i) = &blk.term {
        run(i, stack)
    } else {
        Value::Unit
    }
}

fn run_stacksave<'a, 's, 's3, I>(
    expr: &'a Expr,
    stack: &'s VarStack<'s, Value<'a>>,
    mut stacksave: I,
) -> Value<'a>
where
    I: Iterator<Item = (String, Value<'a>)>,
{
    match stacksave.next() {
        Some((name, value)) => run_stacksave(
            expr,
            &VarStack {
                parent: Some(stack),
                name: &name,
                value,
            },
            stacksave,
        ),
        None => run(expr, stack),
    }
}

pub fn run<'a, 's>(expr: &'a Expr, stack: &'s VarStack<'s, Value<'a>>) -> Value<'a> {
    tracing::trace!("expr={:?}", expr);
    tracing::trace!("stack={:?}", stack);
    use gardswag_syntax::ExprKind as Ek;
    match &expr.inner {
        Ek::Let { lhs, rhs, rest } => {
            let v_rhs = run(rhs, stack);
            run_block(
                rest,
                &VarStack {
                    parent: Some(stack),
                    name: &lhs.inner,
                    value: v_rhs,
                },
            )
        }
        Ek::Block(blk) => run_block(blk, stack),
        Ek::If {
            cond,
            then,
            or_else,
        } => {
            let v_cond = match run(cond, stack) {
                Value::Boolean(x) => x,
                x => panic!("invalid if condition: {:?}", x),
            };
            run_block(if v_cond { then } else { or_else }, stack)
        }
        Ek::Lambda { arg, body } => {
            let mut stacksave = std::collections::BTreeMap::new();
            for (k, v) in stack.iter() {
                if stacksave.contains_key(k) || k == arg.inner || !body.inner.is_var_accessed(k) {
                    continue;
                }
                stacksave.insert(k.to_string(), v.clone());
            }

            Value::Lambda {
                argname: &arg.inner,
                f: body,
                stacksave,
            }
        }
        Ek::Call { prim, arg } => {
            let v_arg = run(arg, stack);
            let v_prim = run(prim, stack);
            match v_prim {
                Value::Builtin { f, mut args } if f.argc() > (args.len() + 1) => {
                    args.push(v_arg);
                    Value::Builtin { f, args }
                }
                Value::Builtin { f, mut args } => {
                    args.push(v_arg);
                    assert_eq!(f.argc(), args.len());
                    use Builtin as Bi;
                    match f {
                        Bi::Plus => match (args.get(0).unwrap(), args.get(1).unwrap()) {
                            (Value::Integer(a), Value::Integer(b)) => Value::Integer(*a + *b),
                            _ => panic!("std.plus called with {:?}", args),
                        },
                        Bi::Minus => match (args.get(0).unwrap(), args.get(1).unwrap()) {
                            (Value::Integer(a), Value::Integer(b)) => Value::Integer(*a - *b),
                            _ => panic!("std.minus called with {:?}", args),
                        },
                        Bi::Mult => match (args.get(0).unwrap(), args.get(1).unwrap()) {
                            (Value::Integer(a), Value::Integer(b)) => Value::Integer(*a * *b),
                            _ => panic!("std.minus called with {:?}", args),
                        },
                        Bi::Leq => match (args.get(0).unwrap(), args.get(1).unwrap()) {
                            (Value::Integer(a), Value::Integer(b)) => Value::Boolean(*a <= *b),
                            _ => panic!("std.minus called with {:?}", args),
                        },
                        Bi::Eq => Value::Boolean(args.get(0) == args.get(1)),
                        Bi::Not => match args.get(0).unwrap() {
                            Value::Boolean(b) => Value::Boolean(!*b),
                            a => panic!("std.not called with {:?}", a),
                        },
                        Bi::StdioWrite => {
                            match args.get(0).unwrap() {
                                Value::PureString(s) => print!("{}", s),
                                x => panic!("std.stdio.write called with {:?}", x),
                            }
                            Value::Unit
                        }
                    }
                }
                Value::Lambda {
                    argname,
                    f,
                    stacksave,
                } => run_stacksave(
                    f,
                    &VarStack {
                        parent: Some(stack),
                        name: argname,
                        value: v_arg,
                    },
                    stacksave.into_iter(),
                ),
                f => panic!("called non-callable {:?} with argument {:?}", f, v_arg),
            }
        }
        Ek::Dot { prim, key } => match run(prim, stack) {
            Value::Record(mut rcm) => rcm
                .remove(&key.inner)
                .expect("unable to find key in record"),
            x => panic!("called .{} on non-record {:?}", key.inner, x),
        },
        Ek::Fix(expr) => {
            run(expr, stack);
            let top = run(expr, stack);
            match &top {
                Value::Lambda {
                    argname,
                    f,
                    stacksave,
                } => {
                    let argname = argname.to_string();
                    let stacksave = stacksave.clone();
                    let f = *f;
                    run_stacksave(
                        f,
                        &VarStack {
                            parent: Some(stack),
                            name: &argname,
                            value: top,
                        },
                        stacksave.into_iter(),
                    )
                }
                f => panic!("called non-callable {:?} with itself as argument", f),
            }
        }
        Ek::FormatString(fsts) => {
            let mut r = String::new();
            for i in fsts {
                r += &match run(i, stack) {
                    Value::PureString(s) => s,
                    Value::Integer(i) => i.to_string(),
                    Value::Boolean(b) => format!("_{}", if b { '1' } else { '0' }),
                    Value::Unit => String::new(),
                    x => panic!("invoked format' stringify on non-stringifyable {:?}", x),
                };
            }
            Value::PureString(r)
        }
        Ek::Record(rcde) => {
            let mut rcd = BTreeMap::new();
            for (k, v) in rcde {
                rcd.insert(k.clone(), run(v, stack));
            }
            Value::Record(rcd)
        }
        Ek::Identifier(id) => stack.find(&id.inner).unwrap().clone(),
        Ek::Boolean(b) => Value::Boolean(*b),
        Ek::Integer(i) => Value::Integer(*i),
        Ek::PureString(s) => Value::PureString(s.clone()),
    }
}
