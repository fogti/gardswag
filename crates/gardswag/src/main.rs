use std::path::PathBuf;
use tracing::debug;

mod infer;
mod interp;

#[derive(Clone, PartialEq, clap::ArgEnum)]
enum Mode {
    /// check if the file passes parsing + type-check
    Check,

    /// perform checks, then execute the expression
    Run,
}

#[derive(clap::Parser)]
#[clap(version)]
struct Args {
    /// file to interpret
    #[clap(short, long)]
    file: PathBuf,

    /// specify the mode
    #[clap(arg_enum, short = 'm', long)]
    mode: Mode,
}

fn mk_env_std(ctx: &mut gardswag_typesys::Context) -> gardswag_typesys::Scheme {
    use gardswag_typesys::{Scheme as TyScheme, Ty, TyLit};

    macro_rules! tl {
        (int) => {{
            Ty::Literal(TyLit::Int)
        }};
        (bool) => {{
            Ty::Literal(TyLit::Bool)
        }};
        (str) => {{
            Ty::Literal(TyLit::String)
        }};
        (()) => {{
            Ty::Literal(TyLit::Unit)
        }};
    }
    macro_rules! ta {
        ($a:expr $(=> $b:expr)*) => {{
            let mut x = vec![$a];
            $(x.push($b);)+
            let mut it = x.into_iter().rev();
            let mut y = it.next().unwrap();
            for i in it {
                y = Ty::Arrow(Box::new(i), Box::new(y));
            }
            y
        }}
    }
    macro_rules! tr {
        ({ $($key:ident: $value:expr),* $(,)? }) => {{
            Ty::Record([
                $((stringify!($key).to_string(), $value)),*
            ].into_iter().collect())
        }};
    }

    let tyvars: Vec<usize> = (0..2).map(|_| ctx.fresh_tyvars.next().unwrap()).collect();
    let gtv = |i| Ty::Var(tyvars[i]);

    let t = tr!({
        plus: ta!(tl!(int) => tl!(int) => tl!(int)),
        minus: ta!(tl!(int) => tl!(int) => tl!(int)),
        mult: ta!(tl!(int) => tl!(int) => tl!(int)),
        eq: ta!(gtv(0) => gtv(1) => tl!(bool)),
        leq: ta!(tl!(int) => tl!(int) => tl!(bool)),
        not: ta!(tl!(bool) => tl!(bool)),
        stdio: tr!({
            write: ta!(tl!(str) => tl!(())),
        }),
    });

    TyScheme {
        forall: tyvars
            .into_iter()
            .map(|_| (ctx.fresh_tyvars.next().unwrap(), Default::default()))
            .collect(),
        t,
    }
}

fn main_check(dat: &str) -> (gardswag_syntax::Block, gardswag_typesys::Scheme) {
    let parsed = gardswag_syntax::parse(&dat).expect("unable to parse file");

    let mut ctx = gardswag_typesys::Context::default();

    let mut env = infer::Env {
        vars: [("std".to_string(), mk_env_std(&mut ctx))]
            .into_iter()
            .collect(),
    };

    match infer::infer_block(&env, &mut ctx, &parsed) {
        Ok(t) => {
            use gardswag_typesys::Substitutable as _;
            debug!("type check ok");
            debug!("=T> {}", t);
            // generalize the type
            env.vars.apply(&ctx);
            let tg = t.clone().generalize(&env, &ctx);
            // garbage collection
            env.gc(&mut ctx, core::iter::once(t));
            debug!("--TV--");
            for (k, v) in &ctx.m {
                debug!("\t${}:\t{}", k, v);
            }
            debug!("--CG--");
            for (k, v) in &ctx.g {
                debug!("\t${}:\t{:?}", k, v);
            }
            tracing::info!("=G> {}", tg);
            (parsed, tg)
        }
        Err(e) => panic!("type checking error: {:?}", e),
    }
}

fn main_interp(parsed: &gardswag_syntax::Block) -> interp::Value<'_> {
    use interp::{Builtin as Bi, Value as Val};

    let stack = gardswag_varstack::VarStack {
        parent: None,
        name: "std",
        value: Val::Record(
            [
                ("plus", Bi::Plus.into()),
                ("minus", Bi::Minus.into()),
                ("mult", Bi::Mult.into()),
                ("eq", Bi::Eq.into()),
                ("leq", Bi::Leq.into()),
                ("not", Bi::Not.into()),
                (
                    "stdio",
                    Val::Record(
                        [("write", Bi::StdioWrite.into())]
                            .into_iter()
                            .map(|(i, j)| (i.to_string(), j))
                            .collect(),
                    ),
                ),
            ]
            .into_iter()
            .map(|(i, j)| (i.to_string(), j))
            .collect(),
        ),
    };

    interp::run_block(&parsed, &stack)
}

fn main() {
    let args = <Args as clap::Parser>::parse();

    let dat = std::fs::read(&args.file).expect("unable to read file");
    let dat = String::from_utf8(dat).expect("file doesn't contain UTF-8 text");

    tracing_subscriber::fmt::init();

    let (parsed, _) = main_check(&dat);

    if args.mode == Mode::Check {
        return;
    }

    let result = main_interp(&parsed);
    println!("result: {:?}", result);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn chk_hello() {
        insta::assert_yaml_snapshot!(main_check(r#"std.stdio.write("Hello world!\n");"#));
    }

    #[test]
    fn chk_fibo() {
        insta::assert_yaml_snapshot!(main_check(
            r#"
                let rec fib = \x \y \n {
                  (* seq: [..., x, y] ++ [z] *)
                  let z = std.plus x y;
                  if (std.leq n 0)
                    { z }
                    { fib y z (std.minus n 1) }
                };
                fib
            "#
        ));
    }

    #[test]
    fn run_fibo0() {
        let x = main_check(
            r#"
                let rec fib = \x \y \n {
                  (* seq: [..., x, y] ++ [z] *)
                  let z = std.plus x y;
                  if (std.leq n 0)
                    { z }
                    { fib y z (std.minus n 1) }
                };
                fib 1 1 0
            "#,
        );
        insta::assert_yaml_snapshot!(main_interp(&x.0));
    }


    #[test]
    fn run_fibo1() {
        let x = main_check(
            r#"
                let rec fib = \x \y \n {
                  (* seq: [..., x, y] ++ [z] *)
                  let z = std.plus x y;
                  if (std.leq n 0)
                    { z }
                    { fib y z (std.minus n 1) }
                };
                fib 1 1 1
            "#,
        );
        insta::assert_yaml_snapshot!(main_interp(&x.0));
    }

    #[test]
    fn run_fibo() {
        let x = main_check(
            r#"
                let rec fib = \x \y \n {
                  (* seq: [..., x, y] ++ [z] *)
                  let z = std.plus x y;
                  if (std.leq n 0)
                    { z }
                    { fib y z (std.minus n 1) }
                };
                fib 1 1 5
            "#,
        );
        insta::assert_yaml_snapshot!(x);
        insta::assert_yaml_snapshot!(main_interp(&x.0));
    }

    #[test]
    fn chk_implicit_restr() {
        insta::assert_yaml_snapshot!(main_check(
            r#"
                \x
                let id = \y y;
                .{
                  id;
                  x;
                  y = "{x}";
                }
            "#
        ));
    }

    #[test]
    fn run_id() {
        let x = main_check(
            r#"
                let id = \x x;
                id 1
            "#,
        );
        insta::assert_yaml_snapshot!(x);
        insta::assert_yaml_snapshot!(main_interp(&x.0));
    }

    #[test]
    fn run_call_blti() {
        let x = main_check(
            r#"
                std.plus 1 1
            "#,
        );
        insta::assert_yaml_snapshot!(x);
        insta::assert_yaml_snapshot!(main_interp(&x.0));
    }
}
