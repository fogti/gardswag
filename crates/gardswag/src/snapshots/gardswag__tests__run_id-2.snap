---
source: crates/gardswag/src/main.rs
assertion_line: 243
expression: main_interp(&x.0)
---
Integer: 1

